server/auth/google.js

```diff
@@ -1,57 +1,55 @@
 // @flow
-import Sequelize from 'sequelize';
 import crypto from 'crypto';
-import Router from 'koa-router';
+import Router, { type Context } from 'koa-router';
+import fetch from 'isomorphic-fetch';
+import { mountOAuth2Passport, type DeserializedData } from '../utils/passport';
 import { capitalize } from 'lodash';
-import { OAuth2Client } from 'google-auth-library';
-import { User, Team, Event } from '../models';
-import auth from '../middlewares/authentication';
-
-const Op = Sequelize.Op;
+import { customError } from '../errors';
+
+class GoogleHDError extends customError('GoogleHDError', 'google-hd') {}
+class HDNotAllowedError extends customError(
+  'HDNotAllowedError',
+  'hd-not-allowed'
+) {}
+
+async function json<T>(
+  input: string | Request | URL,
+  init?: RequestOptions
+): Promise<T> {
+  const res = await fetch(input, init);
+  return await res.json();
+}
+
+async function handleAuthorizeFailed(ctx: Context, err: Error) {
+  switch (true) {
+    case err instanceof GoogleHDError:
+    case err instanceof HDNotAllowedError:
+      ctx.redirect(`/?notice=${err.name}`);
+      break;
+    default:
+      throw err;
+  }
+}
 
-const router = new Router();
-const client = new OAuth2Client(
-  process.env.GOOGLE_CLIENT_ID,
-  process.env.GOOGLE_CLIENT_SECRET,
-  `${process.env.URL}/auth/google.callback`
-);
 const allowedDomainsEnv = process.env.GOOGLE_ALLOWED_DOMAINS;
-
-// start the oauth process and redirect user to Google
-router.get('google', async ctx => {
-  // Generate the url that will be used for the consent dialog.
-  const authorizeUrl = client.generateAuthUrl({
-    access_type: 'offline',
-    scope: [
-      'https://www.googleapis.com/auth/userinfo.profile',
-      'https://www.googleapis.com/auth/userinfo.email',
-    ],
-    prompt: 'consent',
-  });
-  ctx.redirect(authorizeUrl);
-});
-
-// signin callback from Google
-router.get('google.callback', auth({ required: false }), async ctx => {
-  const { code } = ctx.request.query;
-  ctx.assertPresent(code, 'code is required');
-  const response = await client.getToken(code);
-  client.setCredentials(response.tokens);
-
-  const profile = await client.request({
-    url: 'https://www.googleapis.com/oauth2/v1/userinfo',
+async function deserializeGoogleToken(
+  accessToken,
+  refreshToken: string
+): Promise<DeserializedData> {
+  const profile = await json('https://www.googleapis.com/oauth2/v1/userinfo', {
+    headers: {
+      authorization: `Bearer ${accessToken}`,
+    },
   });
 
   if (!profile.data.hd) {
```

Comment:
GoogleHDError is never thrown when trying to login with a regular Google account, since the check for the condition cannot be performed, as it just happened to me.

```
TypeError: Cannot read property 'hd' of undefined
    at _callee3$ (/opt/outline/server/auth/google.js:45:21)
    at tryCatch (/opt/outline/node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:65:40)
    at Generator.invoke [as _invoke] (/opt/outline/node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:303:22)
    at Generator.prototype.<computed> [as next] (/opt/outline/node_modules/babel-polyfill/node_modules/regenerator-runtime/runtime.js:117:21)
    at step (/opt/outline/server/auth/google.js:177:191)
    at /opt/outline/server/auth/google.js:177:361
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```